src/common/upload.ts

```diff
@@ -118,36 +125,80 @@ export abstract class TurboAuthenticatedBaseUploadService
-        fileStreamFactory,
-        fileSizeFactory,
-        dataItemOpts,
-      });
-    const signedDataItem = dataItemStreamFactory();
-    this.logger.debug('Uploading signed data item...');
-    // TODO: add p-limit constraint or replace with separate upload class
+    let retries = 0;
+    const maxRetries = this.retryConfig.retries ?? 3;
+    const retryDelay =
+      this.retryConfig.retryDelay ??
+      ((retryNumber: number) => retryNumber * 1000);
+    let lastError = 'Unknown Error'; // Store the last error for throwing
+    let lastStatusCode: number | undefined; // Store the last status code for throwing
+
+    while (retries < maxRetries) {
+      if (signal?.aborted) {
+        throw new CanceledError();
+      }
 
-    const headers = {
-      'content-type': 'application/octet-stream',
-      'content-length': `${dataItemSizeFactory()}`,
-    };
-    if (dataItemOpts !== undefined && dataItemOpts.paidBy !== undefined) {
-      const paidBy = Array.isArray(dataItemOpts.paidBy)
-        ? dataItemOpts.paidBy
-        : [dataItemOpts.paidBy];
+      const { dataItemStreamFactory, dataItemSizeFactory } =
+        await this.signer.signDataItem({
+          fileStreamFactory,
+          fileSizeFactory,
+          dataItemOpts,
+        });
+
+      try {
+        this.logger.debug('Uploading signed data item...');
+        // TODO: add p-limit constraint or replace with separate upload class
+
+        const headers = {
+          'content-type': 'application/octet-stream',
+          'content-length': `${dataItemSizeFactory()}`,
+        };
+        if (dataItemOpts !== undefined && dataItemOpts.paidBy !== undefined) {
+          const paidBy = Array.isArray(dataItemOpts.paidBy)
+            ? dataItemOpts.paidBy
+            : [dataItemOpts.paidBy];
+
+          if (dataItemOpts.paidBy.length > 0) {
+            headers['x-paid-by'] = paidBy;
+          }
+        }
+        const data = await this.httpService.post<TurboUploadDataItemResponse>({
+          endpoint: `/tx/${this.token}`,
+          signal,
+          data: dataItemStreamFactory(),
+          headers,
+        });
+        return data;
+      } catch (error) {
+        if (error instanceof FailedRequestError) {
+          throw error; // Rethrow if it's already a failed request error from HTTP service
+        }
 
-      if (dataItemOpts.paidBy.length > 0) {
-        headers['x-paid-by'] = paidBy;
+        // Store the last encountered error and status for re-throwing after retries
+        if (error instanceof AxiosError) {
+          lastStatusCode = error.response?.status;
+          lastError = error.code ?? error.message;
+        } else {
+          lastError =
+            error instanceof Error && error.message !== undefined
+              ? error.message
+              : `${error}`; // Stringify the whole error if it's not a known Error instance
+        }
+        this.logger.debug(
+          `Upload failed, attempt ${retries + 1}/${maxRetries}`,
```

Comment:
Hmmm attempts is different than retries. If you allow for 3 retries, that's a total of 4 attempts. So you can either:
- Reimagine `retries` as `tries` and `max retries` as `max tries`
- change your log to `maxRetries + 1` (e.g. Attempt 1/4)
- update your log to say something like `Upload failed. Will attempt retry 1/3...`

Reply:
changed to the second option there using `Attempts` in this commit: c4c2617

so `(retries + 1)/(maxRetries + 1 )`